src/connectors/hits/connectHits.ts

```diff
@@ -6,55 +6,62 @@ import {
   addQueryID,
   noop,
 } from '../../lib/utils';
+import {
+  RendererOptions,
+  WidgetFactory,
+  Unmounter,
+  Renderer,
+} from '../../types';
+import { SearchResults } from 'algoliasearch-helper';
 
 const withUsage = createDocumentationMessageGenerator({
   name: 'hits',
   connector: true,
 });
 
-/**
- * @typedef {Object} HitsRenderingOptions
- * @property {Object[]} hits The matched hits from Algolia API.
- * @property {Object} results The complete results response from Algolia API.
- * @property {Object} widgetParams All original widget options forwarded to the `renderFn`.
- */
-
-/**
- * @typedef {Object} CustomHitsWidgetOptions
- * @property {boolean} [escapeHTML = true] Whether to escape HTML tags from `hits[i]._highlightResult`.
- * @property {function(Object[]):Object[]} [transformItems] Function to transform the items passed to the templates.
- */
-
-/**
- * **Hits** connector provides the logic to create custom widgets that will render the results retrieved from Algolia.
- * @type {Connector}
- * @param {function(HitsRenderingOptions, boolean)} renderFn Rendering function for the custom **Hits** widget.
- * @param {function} unmountFn Unmount function called when the widget is disposed.
- * @return {function(CustomHitsWidgetOptions)} Re-usable widget factory for a custom **Hits** widget.
- * @example
- * // custom `renderFn` to render the custom Hits widget
- * function renderFn(HitsRenderingOptions) {
- *   HitsRenderingOptions.widgetParams.containerNode.html(
- *     HitsRenderingOptions.hits.map(function(hit) {
- *       return '<div>' + hit._highlightResult.name.value + '</div>';
- *     })
- *   );
- * }
- *
- * // connect `renderFn` to Hits logic
- * var customHits = instantsearch.connectors.connectHits(renderFn);
- *
- * // mount widget on the page
- * search.addWidgets([
- *   customHits({
- *     containerNode: $('#custom-hits-container'),
- *   })
- * ]);
- */
-export default function connectHits(renderFn, unmountFn = noop) {
+export type HitsRendererOptions<THitsWidgetParams> = {
+  /**
+   * The matched hits from Algolia API.
+   */
+  hits: Array<Record<string, any>>;
+
+  /**
+   * The response from Algolia API.
+   */
+  results?: SearchResults<any>;
+} & RendererOptions<THitsWidgetParams & HitsConnectorParams>;
+
+export type HitsConnectorParams = {
+  /**
+   * Whether to escape HTML tags from `hits[i]._highlightResult`.
+   *
+   * @default true
+   */
+  escapeHTML?: boolean;
+
+  /**
+   * Function to transform the items passed to the templates.
+   */
+  transformItems?: (
+    objects: Array<Record<string, any>>
+  ) => Array<Record<string, any>>;
+};
+
+export type HitsRenderer<THitsWidgetParams> = Renderer<
+  HitsRendererOptions<THitsWidgetParams & HitsConnectorParams>
+>;
+
+export type HitsWidgetFactory<THitsWidgetParams> = WidgetFactory<
+  THitsWidgetParams & HitsConnectorParams
+>;
+
+export default function connectHits<THitsWidgetParams = {}>(
```

Comment:
Should we default `THitsWidgetParams` to `any` here? Right now, users _have to_ specify a type when using the widget params.

Reply:
We are **defaulting to empty generic**, so they are not forced.

if they don't pass a generic, the params will be {} & connectorParams.

Reply:
it says `= {}`, no?

Reply:
I meant, if you want to access a widget param (e.g. `container`), you need to specify the `connectHits` type:

https://github.com/algolia/instantsearch.js/blob/62afb953f1bd34b53727eb6f2c6e37ae26e72596/stories/configure-related-items.stories.ts#L71-L83

Wondering if we should be less strict about it by defaulting to `any`?

Reply:
@francoischalifour I don't think is a good idea. The goal of using TypeScript is provide type safety.  Defaulting to any, is defaulting to non safe code.

Reply:
@Haroenv If you do `connectHits(` without generic, means that you don't have any widget params, besides the default connector params ones.

Reply:
Makes sense!